modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java

```diff
@@ -343,4 +360,171 @@ static boolean hasSubQuery(SqlNode node) {
             return super.visit(call);
         }
     }
+
+
+    /**
+     * Tries to optimize a query that looks like {@code SELECT count(*)}.
+     *
+     * @param planner Planner.
+     * @param txContext Transactional context.
+     * @param node Query node.
+     * @return Plan node with list of aliases, if the optimization is applicable.
+     */
+    public static @Nullable Pair<IgniteRel, List<String>> tryOptimizeSelectCount(
+            IgnitePlanner planner,
+            @Nullable QueryTransactionContext txContext,
+            SqlNode node
+    ) {
+        if (txContext != null && txContext.explicitTx() != null) {
+            return null;
+        }
+
+        if (!(node instanceof SqlSelect)) {
+            return null;
+        }
+
+        SqlSelect select = (SqlSelect) node;
+
+        if (select.getGroup() != null
+                || select.getWhere() != null
+                || select.getHaving() != null
+                || select.getQualify() != null
+                || !select.getWindowList().isEmpty()
+                || select.getOffset() != null
+                || select.getFetch() != null) {
+            return null;
+        }
+
+        IgniteSqlToRelConvertor converter = planner.sqlToRelConverter();
+
+        RelOptTable targetTable;
+
+        if (select.getFrom() != null) {
+            SqlNode from = SqlUtil.stripAs(select.getFrom());
+            // Skip non-references such as VALUES ..
+            if (from.getKind() != SqlKind.IDENTIFIER) {
+                return null;
+            }
+
+            targetTable = converter.getTargetTable(select.getFrom());
+            IgniteDataSource dataSource = targetTable.unwrap(IgniteDataSource.class);
+            if (!(dataSource instanceof IgniteTable)) {
+                return null;
+            }
+        } else {
+            // Try to optimize SELECT count(*) as well
```

Comment:
what is the point of such optimization? What is the difference between new path and old path in numbers?

Reply:
Removed this piece of code.